
<!-- README.md is generated from README.Rmd. Please edit that file -->

(This is a draft)

# variantCallingFromIsoSeq

<!-- badges: start -->
<!-- badges: end -->

Use ground-truth VCF files generated from short-read data to compare and
validate VCF files generated from different variant callers on Iso-Seq
data.

## Installation

To install it use
`devtools::install_github("vladimirsouza/variantCallingFromIsoSeq@main")`.

## An example on how to construct a master table

We construct a master table that contains information about the variants
called by the methods to be validated and the variants in the
ground-truth. From this master table, we can generate plots and analyze
and compare the VCF files used to construct the table.

### Introduction

Generate a new master table to compare and validate the variant calls
from Iso-Seq data when using DeepVariant alone and SplitNCigarReads +
DeepVariant. The ground-truth was generated by GATK pipeline on
short-read DNA data.

### Input variables

``` r
### methods to validate
# name of the methods to validate
METHOD_NAMES <- c("dv", "dv_s")
# name of the dataset used with the methods to validate
METHOD_DATASET_NAME <- "isoSeq"
# VCF files
METHOD1_VCF_FILE <- "/home/vbarbo/project_2021/datasets/gloria_data/analysis/dv_calls/noSplitBam/deepvariant_calls_pass.vcf.gz"
METHOD2_VCF_FILE <- "/home/vbarbo/project_2021/datasets/gloria_data/analysis/dv_calls/deepvariant_calls_pass.vcf.gz"
# BAM of the data
METHOD_BAM_FILE <- "/home/vbarbo/project_2021/datasets/gloria_data/analysis/dv_calls/aln_rg_dedupped.bam"


### ground-truth
# name
TRUTH_NAME <- "tr_dna_merged"
# name of the dataset used to generate the ground-truth
TRUTH_DATASET_NAME <- "shortRead"
# VCF file
TRUTH_VCF_FILE <- "/home/vbarbo/project_2021/datasets/gloria_data/analysis/truth_merged_bams/merged.recal_exons_pass.vcf.gz"
# BAM of the data
TRUTH_BAM_FILE <- "/home/vbarbo/project_2021/datasets/gloria_data/analysis/truth_merged_bams/merged.bam"


### variables
MAX_DIST_FROM_SPLICE_SITE <- 20
THREADS <- 40
```

### libraries

``` r
library(variantCallingFromIsoSeq)
library(GenomicAlignments)
library(dplyr)
```

### Create the master table

#### Initiate the master table

Get variant positions and which method could call them.

``` r
dat <- initiate_master_table(
  METHOD1_VCF_FILE, 
  METHOD2_VCF_FILE, 
  TRUTH_VCF_FILE,
  method_names=c(METHOD_NAMES, TRUTH_NAME)
)
```

#### get splice site positions from the BAM file

NOTE: This function may take a long time to run.

``` r
### don't run this
splice_sites <- get_splice_sites_info(input_bam, THREADS)
```

#### Add columns about splice sites

``` r
dat <- add_splice_site_info_to_master_table(
  dat, splice_sites,
  MAX_DIST_FROM_SPLICE_SITE
)
```

#### Add the read coverage (from the BAM file) of each variant

First, we need to load the BAM file and get the coverage with
`IRanges::coverage`.

``` r
method_bam <- readGAlignments(METHOD_BAM_FILE)
method_coverage <- coverage(method_bam)

truth_bam <- readGAlignments(TRUTH_BAM_FILE)
truth_coverage <- coverage(truth_bam)
```

Add read coverage of each dataset.

``` r
dat <- add_read_coverage_from_bam_to_master_table(
  method_coverage,
  truth_coverage,
  input_table=dat,
  dataset_names=c(METHOD_DATASET_NAME, TRUTH_DATASET_NAME)
)
```

#### Add the number of N-cigar reads per site

``` r
dat <- add_number_of_n_cigar_reads_to_master_table(
  dat, 
  method_bam, 
  METHOD_DATASET_NAME
)
```

#### Add column to comprare two methods

Compare `dv` and `dv_s`. Do they call the same varaints?

``` r
dat <- add_two_method_comparison_to_master_table(
  dat, 
  METHOD_NAMES[1], 
  METHOD_NAMES[2]
)
```

#### Add column to classify method calls (compare to the ground-truth)

Classify `dv` calls.

``` r
dat <- add_method_vs_truth_comparison_to_master_table(
  dat, 
  METHOD_NAMES[1], 
  TRUTH_NAME
)
```

Classify `dv_s` calls.

``` r
dat <- add_method_vs_truth_comparison_to_master_table(
  dat, 
  METHOD_NAMES[2], 
  TRUTH_NAME
)
```

### Take a look at the master table

``` r
head(dat)
#>   chrm   pos in_dv in_dv_s in_tr_dna_merged dp_dv dp_dv_s dp_tr_dna_merged is_near_ss ss_dist
#> 1 chr1 15274     1       1                0     5       2               NA          0      NA
#> 2 chr1 15817     1       0                0    19      NA               NA          0      NA
#> 3 chr1 15820     1       0                0    19      NA               NA          0      NA
#> 4 chr1 15903     1       0                0    19      NA               NA          0      NA
#> 5 chr1 16933     1       0                0    20      NA               NA          0      NA
#> 6 chr1 18849     1       0                0     2      NA               NA          0      NA
#>   is_acceptor_site ss_num is_single_ss ss_shortest_dist ss_highest_num is_acceptor_site_mode
#> 1               NA     NA           -1             2000              0                  <NA>
#> 2               NA     NA           -1             2000              0                  <NA>
#> 3               NA     NA           -1             2000              0                  <NA>
#> 4               NA     NA           -1             2000              0                  <NA>
#> 5               NA     NA           -1             2000              0                  <NA>
#> 6               NA     NA           -1             2000              0                  <NA>
#>   isoSeq_coverage shortRead_coverage isoSeq_ncr_num compare_dv_dvS dv_classification
#> 1              16                  0             69           both                FP
#> 2              83                  3              0             dv                FP
#> 3              83                  3              0             dv                FP
#> 4              83                  0              0             dv                FP
#> 5              89                  5              0             dv                FP
#> 6               4                  0             81             dv                FP
#>   dv_s_classification
#> 1                  FP
#> 2                  TN
#> 3                  TN
#> 4                  TN
#> 5                  TN
#> 6                  TN
```

### Subset and calculate the precision, sensitivity and F1-score of the methods

Take only variants in well-coverage regions in both Iso-Seq and
short-read data.

``` r
well_coverage_dat <- filter(dat, 
                            isoSeq_coverage>=10 & shortRead_coverage>=40 &
                              shortRead_coverage<=70)
```

Calculate accuracy measures for dv.

``` r
calc_precision_sensitivity_f1Score(well_coverage_dat, METHOD_NAMES[1], TRUTH_NAME)
#>   precision sensitivity     f1Score 
#>   0.8005915   0.6610849   0.7241808
```

Calculate accuracy measures for dv\_s.

``` r
calc_precision_sensitivity_f1Score(well_coverage_dat, METHOD_NAMES[2], TRUTH_NAME)
#>   precision sensitivity     f1Score 
#>   0.7836834   0.6442526   0.7071606
```
