# This script is to identify variant sites that come from ASE genes. 
# To do that, an input table, generated by GATK ASEReadCounter, of read
# counts for each allele is used.
# the chi-squared test to identify ASE sites is used.




### inputs
allele_counts_file <- "/home/vbarbo/project_2021/datasets/wtc11/ase_analysis/GSE175048_ENCSR673UKZ/ase_read_count_tables/ase_read_count_all.table"


### packages
library(dplyr)
library(readr)


### load allele counts
allele_counts <- read_tsv(allele_counts_file)


### add column of p-values for ase identification
ase_chiSquare <- mapply(function(ref, alt){
  suppressWarnings(
    tryCatch(
      chisq.test( c(ref, alt) ) $p.value,
      error=function(e){NA}
    )
  )
}, allele_counts$refCount, allele_counts$altCount)

ase_chiSquare_adj <- p.adjust(ase_chiSquare, method="BH")

allele_counts <- mutate(allele_counts, ase_chiSquare_adj)


### add ase q-value to the master table
allele_counts <- rename(allele_counts, "chrm"="contig", "pos"="position")
allele_p <- subset(allele_counts,
                   select=c("chrm", "pos", "refCount", "altCount",
                            "totalCount", "otherBases", "ase_chiSquare_adj"))

load("~/project_2021/scripts/paper_1/master_tables/wtc11/mt_wtc11_allMethods_filtered_v5.RData")
dat <- mt_wtc11_allMethods_filtered
dat1 <- merge(dat, allele_p, by=c("chrm", "pos"), all.x=TRUE, sort=FALSE)


### keep only variants that are SNPs and heterozygous, according to the 
### ground truth, and have a p-value for ASE identification associated to 
### them, i.e. are also SNPs and heterozygous according to the rna short reads
dat2 <- filter(dat1, variantType_allen=="snp" & gt_allen=="0/1")
dat2 <- filter(dat2, !is.na(ase_chiSquare_adj))


### keep only variants with reasonable read coverage in the ground truth and
### in the rna short read data
dat2 <- filter(dat2, isoSeq_coverage>=20)
dat2 <- filter(dat2, totalCount>=40)


### chi-square test to test whether the proportions of FNs between ASE and 
### non-ASE variants are different
k <- table(dat2$dv_s_fc_classification, dat2$ase_chiSquare_adj<.05)
k <- k[rownames(k) %in% c("FN", "TP"),]
chisq.test(k)$p.value  # 1.465621e-42



### create the tables from which ggplot2 make the plot

# sncr + flagcorrecion + deepvariant
datf <- table(dat2$dv_s_fc_classification, dat2$ase_chiSquare_adj<.05)
datf <- datf[c("FN", "TP"),]
datf <- as.data.frame(datf)
names(datf) <- c("Classification", "ASE_Identification", "Frequency")
datf$ASE_Identification <- ifelse( as.logical(datf$ASE_Identification),
                                   "ASE", "Non ASE") %>% factor
datf$Ratio <- 0
datf$Ratio[1] <- datf$Freq[1] / sum(datf$Freq[1:2])
datf$Ratio[2] <- datf$Freq[2] / sum(datf$Freq[1:2])
datf$Ratio[3] <- datf$Freq[3] / sum(datf$Freq[3:4])
datf$Ratio[4] <- datf$Freq[4] / sum(datf$Freq[3:4])

datf_dvSFc <- datf


# sncr + gatk
datf <- table(dat2$gatk_s_classification, dat2$ase_chiSquare_adj<.05)
datf <- datf[c("FN", "TP"),]
datf <- as.data.frame(datf)
names(datf) <- c("Classification", "ASE_Identification", "Frequency")
datf$ASE_Identification <- ifelse( as.logical(datf$ASE_Identification),
                                   "ASE", "Non ASE") %>% factor
datf$Ratio <- 0
datf$Ratio[1] <- datf$Freq[1] / sum(datf$Freq[1:2])
datf$Ratio[2] <- datf$Freq[2] / sum(datf$Freq[1:2])
datf$Ratio[3] <- datf$Freq[3] / sum(datf$Freq[3:4])
datf$Ratio[4] <- datf$Freq[4] / sum(datf$Freq[3:4])

datf_gatkS <- datf


# clair mix
datf <- table(dat2$c3_mix_classification, dat2$ase_chiSquare_adj<.05)
datf <- datf[c("FN", "TP"),]
datf <- as.data.frame(datf)
names(datf) <- c("Classification", "ASE_Identification", "Frequency")
datf$ASE_Identification <- ifelse( as.logical(datf$ASE_Identification),
                                   "ASE", "Non ASE") %>% factor
datf$Ratio <- 0
datf$Ratio[1] <- datf$Freq[1] / sum(datf$Freq[1:2])
datf$Ratio[2] <- datf$Freq[2] / sum(datf$Freq[1:2])
datf$Ratio[3] <- datf$Freq[3] / sum(datf$Freq[3:4])
datf$Ratio[4] <- datf$Freq[4] / sum(datf$Freq[3:4])

datf_c3Mix <- datf


### put all tables together in a single table


datf_dvSFc$method <- "SNCR+FC+DeepVariant"
datf_c3Mix$method <- "Clair3 mix"
datf_gatkS$method <- "SNCR+GATK"
datf <- rbind(datf_dvSFc, datf_c3Mix, datf_gatkS)
k <- c("SNCR+FC+DeepVariant", "Clair3 mix", "SNCR+GATK")
datf$method <- factor(datf$method, levels=k, ordered=TRUE)

save(datf, file="~/load_later/ase_analysis/all_methods_info_to_plot.RData")

